utils/codec/manager.go

```diff
@@ -0,0 +1,127 @@
+// (c) 2019-2020, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package codec
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/utils/wrappers"
+)
+
+const (
+	// default max size, in bytes, of something being marshalled by Marshal()
+	defaultMaxSize = 1 << 18
+
+	// initial capacity of byte slice that values are marshaled into.
+	// Larger value --> need less memory allocations but possibly have allocated but unused memory
+	// Smaller value --> need more memory allocations but more efficient use of allocated memory
+	initialSliceCap = 128
+)
+
+var (
+	errUnknownVersion    = errors.New("unknown codec version")
+	errDuplicatedVersion = errors.New("duplicated codec version")
+)
+
+// Manager describes the functionality for managing codec versions.
+type Manager interface {
+	RegisterCodec(version uint16, codec Codec) error
+	SetMaxSize(int)
+	Marshal(version uint16, source interface{}) (destination []byte, err error)
+	Unmarshal(source []byte, destination interface{}) (version uint16, err error)
+}
+
+// NewNewManager returns a new codec manager.
```

Comment:
` // Unmarshal the given bytes into the given destination. [destination] must be a pointer or an interface. Returns the version of the codec that produces the given bytes.`